### Importando arquivos .csv

In [2]:
import pandas as pd

Ler os dados na planilha GDP per capita.

- Selecionamos a planilha do arquivo com os dados que precisamos, pulando linhas e colunas que não queremos. 
- Usamos o parâmetro `sheet_name` para especificar a planilha. 
- Usamos o `skiprows=4` e `skipfooter=1` para pularmos as quatro primeiras linhas (a primeira linha está escondida) e a última linha.
- Damos valores em `usecols` para pegar os dados da **coluna A** e **coluna C até T**. 
- Usamos `head` para vermos as primeiras linhas do dataset.

In [3]:
percapitaGDP = pd.read_excel('data/GDPpercapita.xlsx',
    sheet_name='OECD.Stat export',
    skiprows=4,
    skipfooter=1,
    usecols='A, C:T')

In [4]:
percapitaGDP.head()

Year   2001   2002   2003   2004   2005   2006  \
0          Metropolitan areas    NaN    NaN    NaN    NaN    NaN    NaN   
1              AUS: Australia     ..     ..     ..     ..     ..     ..   
2       AUS01: Greater Sydney  43313  44008  45424  45837  45423  45547   
3    AUS02: Greater Melbourne  40125  40894  41602  42188  41484  41589   
4     AUS03: Greater Brisbane  37580  37564  39080  40762  42976  44475   

    2007   2008   2009   2010   2011   2012   2013   2014   2015   2016  \
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1     ..     ..     ..     ..     ..     ..     ..     ..     ..     ..   
2  45880  45225  45900  45672  46535  47350  47225  48510  50075  50519   
3  42316  40975  41384  40943  41165  41264  41157  42114  42928  42671   
4  44635  46192  43507  42774  44166  43764  43379  43754  44388  45723   

    2017   2018  
0    NaN    NaN  
1     ..     ..  
2  50578  49860  
3  43025  42674  
4  46876  46640

Usamos o método `info` para vermos os tipos dos dados e uma contagem de dados non-null

In [5]:
percapitaGDP.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 19 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Year    702 non-null    object
 1   2001    701 non-null    object
 2   2002    701 non-null    object
 3   2003    701 non-null    object
 4   2004    701 non-null    object
 5   2005    701 non-null    object
 6   2006    701 non-null    object
 7   2007    701 non-null    object
 8   2008    701 non-null    object
 9   2009    701 non-null    object
 10  2010    701 non-null    object
 11  2011    701 non-null    object
 12  2012    701 non-null    object
 13  2013    701 non-null    object
 14  2014    701 non-null    object
 15  2015    701 non-null    object
 16  2016    701 non-null    object
 17  2017    701 non-null    object
 18  2018    701 non-null    object
dtypes: object(19)
memory usage: 104.3+ KB



- Renomeamos a coluna `Year` para `metro` e removemos os espaços do início.
- Existem espaços extras antes e depois dos valores na coluna `metro` em alguns casos. 
- Nós podemos testar se existem tais espaços com o comando `startwith(' ')` e então usar `any` para estabelecer qualquer célula que possua um ou mais ocasiões com o primeiro caracter em branco.
- Da mesma forma, usamos o comando `endswith(' ')` para examinar os espaços depois dos caracteres.
- Usamos `strip` para remover ambos espaços (início e fim dos caracteres)

In [6]:
percapitaGDP.rename(columns={'Year':'metro'}, inplace=True)

In [7]:
percapitaGDP.metro.str.startswith(' ').any()

True

In [8]:
percapitaGDP.metro.str.endswith(' ').any()

True

In [9]:
percapitaGDP.metro = percapitaGDP.metro.str.strip()

Convertendo as colunas de datas para numéricas.

- Iterando sobre todas as colunas de GDP ear (2001-2018) e convertendo o tipo dos dados de `object` para `float`. 

In [11]:
for col in percapitaGDP.columns[1:]:
    percapitaGDP[col] = pd.to_numeric(percapitaGDP[col], errors='coerce')
    percapitaGDP.rename(columns={col:'pcGDP'+col}, inplace=True)

In [12]:
percapitaGDP.head()

metro  pcGDP2001  pcGDP2002  pcGDP2003  pcGDP2004  \
0        Metropolitan areas        NaN        NaN        NaN        NaN   
1            AUS: Australia        NaN        NaN        NaN        NaN   
2     AUS01: Greater Sydney    43313.0    44008.0    45424.0    45837.0   
3  AUS02: Greater Melbourne    40125.0    40894.0    41602.0    42188.0   
4   AUS03: Greater Brisbane    37580.0    37564.0    39080.0    40762.0   

   pcGDP2005  pcGDP2006  pcGDP2007  pcGDP2008  pcGDP2009  pcGDP2010  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2    45423.0    45547.0    45880.0    45225.0    45900.0    45672.0   
3    41484.0    41589.0    42316.0    40975.0    41384.0    40943.0   
4    42976.0    44475.0    44635.0    46192.0    43507.0    42774.0   

   pcGDP2011  pcGDP2012  pcGDP2013  pcGDP2014  pcGDP2015  pcGDP2016  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2    46535.0    47350.0    47225.0    48510.0    50075.0    50519.0   
3    41165.0    41264.0    41157.0    42114.0    42928.0    42671.0   
4    44166.0    43764.0    43379.0    43754.0    44388.0    45723.0   

   pcGDP2017  pcGDP2018  
0        NaN        NaN  
1        NaN        NaN  
2    50578.0    49860.0  
3    43025.0    42674.0  
4    46876.0    46640.0

In [13]:
percapitaGDP.dtypes

metro         object
pcGDP2001    float64
pcGDP2002    float64
pcGDP2003    float64
pcGDP2004    float64
pcGDP2005    float64
pcGDP2006    float64
pcGDP2007    float64
pcGDP2008    float64
pcGDP2009    float64
pcGDP2010    float64
pcGDP2011    float64
pcGDP2012    float64
pcGDP2013    float64
pcGDP2014    float64
pcGDP2015    float64
pcGDP2016    float64
pcGDP2017    float64
pcGDP2018    float64
dtype: object

Usamos o método `describe` para gerar um sumário estatístico para todos os dados numéricos ni dataframe.

In [14]:
percapitaGDP.describe()

pcGDP2001     pcGDP2002     pcGDP2003     pcGDP2004      pcGDP2005  \
count    424.000000    440.000000    440.000000    440.000000     447.000000   
mean   41263.658019  41015.070455  41553.361364  42473.022727   42881.143177   
std    11877.960193  12536.516772  12456.583153  12621.901150   13172.229181   
min    10988.000000  11435.000000  11969.000000  12777.000000   13062.000000   
25%    33139.250000  32636.000000  33284.750000  33864.500000   33735.500000   
50%    39543.500000  39683.500000  40390.500000  41200.500000   41609.000000   
75%    47971.750000  48611.000000  49354.750000  50468.250000   51025.000000   
max    91488.000000  93566.000000  98123.000000  96242.000000  101084.000000   

           pcGDP2006      pcGDP2007      pcGDP2008      pcGDP2009  \
count     447.000000     447.000000     455.000000     471.000000   
mean    43987.762864   44786.760626   44533.958242   42724.316348   
std     13450.431995   13693.732714   14082.871703   13602.723246   
min     13855.000000   13937.000000    2236.000000    2202.000000   
25%     34540.000000   35226.000000   35094.000000   33730.000000   
50%     42929.000000   43461.000000   43287.000000   41250.000000   
75%     52304.000000   53043.500000   53132.000000   50739.500000   
max    121053.000000  122897.000000  120158.000000  114486.000000   

           pcGDP2010      pcGDP2011      pcGDP2012      pcGDP2013  \
count     471.000000     480.000000     480.000000     480.000000   
mean    43433.511677   43946.966667   44075.937500   44302.154167   
std     13896.775080   14018.472002   14170.164166   14251.392256   
min      2227.000000    2363.000000    2572.000000    2700.000000   
25%     34294.500000   34582.750000   34808.750000   35113.750000   
50%     41627.000000   42345.000000   42131.500000   42154.500000   
75%     51428.500000   52568.750000   52569.750000   53087.500000   
max    119658.000000  119965.000000  117348.000000  123709.000000   

           pcGDP2014      pcGDP2015      pcGDP2016      pcGDP2017  \
count     480.000000     480.000000     480.000000     445.000000   
mean    44942.493750   45802.220833   46243.666667   47489.089888   
std     14421.619849   14948.683819   14938.549380   15463.803389   
min      2683.000000    2761.000000    2796.000000    2745.000000   
25%     35766.250000   36128.500000   36584.750000   37316.000000   
50%     42777.500000   43237.500000   43931.500000   45385.000000   
75%     53737.250000   54134.250000   54449.750000   56023.000000   
max    121011.000000  121623.000000  117879.000000  122242.000000   

           pcGDP2018  
count     441.000000  
mean    48032.668934  
std     15719.725615  
min      2832.000000  
25%     37908.000000  
50%     46057.000000  
75%     56638.000000  
max    127468.000000

Removemos linhas onde todos os valores de GDP percapita estão faltando.

- Usamos o parâmetro `subset` do `dropna` para inspecionar todas as colunas, começando pela segunda coluna até a última.
- Usamos `how` para especificar que queremos dropar as linhas se todas as colunas especificadas no `subset` possuem dados faltando.
- Comando `shape` para mostrar o número de linhas e colunas no resultado do dataframe.

In [15]:
percapitaGDP.dropna(subset=percapitaGDP.columns[1:], how='all', inplace=True)

In [16]:
percapitaGDP.describe()

pcGDP2001     pcGDP2002     pcGDP2003     pcGDP2004      pcGDP2005  \
count    424.000000    440.000000    440.000000    440.000000     447.000000   
mean   41263.658019  41015.070455  41553.361364  42473.022727   42881.143177   
std    11877.960193  12536.516772  12456.583153  12621.901150   13172.229181   
min    10988.000000  11435.000000  11969.000000  12777.000000   13062.000000   
25%    33139.250000  32636.000000  33284.750000  33864.500000   33735.500000   
50%    39543.500000  39683.500000  40390.500000  41200.500000   41609.000000   
75%    47971.750000  48611.000000  49354.750000  50468.250000   51025.000000   
max    91488.000000  93566.000000  98123.000000  96242.000000  101084.000000   

           pcGDP2006      pcGDP2007      pcGDP2008      pcGDP2009  \
count     447.000000     447.000000     455.000000     471.000000   
mean    43987.762864   44786.760626   44533.958242   42724.316348   
std     13450.431995   13693.732714   14082.871703   13602.723246   
min     13855.000000   13937.000000    2236.000000    2202.000000   
25%     34540.000000   35226.000000   35094.000000   33730.000000   
50%     42929.000000   43461.000000   43287.000000   41250.000000   
75%     52304.000000   53043.500000   53132.000000   50739.500000   
max    121053.000000  122897.000000  120158.000000  114486.000000   

           pcGDP2010      pcGDP2011      pcGDP2012      pcGDP2013  \
count     471.000000     480.000000     480.000000     480.000000   
mean    43433.511677   43946.966667   44075.937500   44302.154167   
std     13896.775080   14018.472002   14170.164166   14251.392256   
min      2227.000000    2363.000000    2572.000000    2700.000000   
25%     34294.500000   34582.750000   34808.750000   35113.750000   
50%     41627.000000   42345.000000   42131.500000   42154.500000   
75%     51428.500000   52568.750000   52569.750000   53087.500000   
max    119658.000000  119965.000000  117348.000000  123709.000000   

           pcGDP2014      pcGDP2015      pcGDP2016      pcGDP2017  \
count     480.000000     480.000000     480.000000     445.000000   
mean    44942.493750   45802.220833   46243.666667   47489.089888   
std     14421.619849   14948.683819   14938.549380   15463.803389   
min      2683.000000    2761.000000    2796.000000    2745.000000   
25%     35766.250000   36128.500000   36584.750000   37316.000000   
50%     42777.500000   43237.500000   43931.500000   45385.000000   
75%     53737.250000   54134.250000   54449.750000   56023.000000   
max    121011.000000  121623.000000  117879.000000  122242.000000   

           pcGDP2018  
count     441.000000  
mean    48032.668934  
std     15719.725615  
min      2832.000000  
25%     37908.000000  
50%     46057.000000  
75%     56638.000000  
max    127468.000000

In [17]:
percapitaGDP.head()

metro  pcGDP2001  pcGDP2002  pcGDP2003  pcGDP2004  \
2     AUS01: Greater Sydney    43313.0    44008.0    45424.0    45837.0   
3  AUS02: Greater Melbourne    40125.0    40894.0    41602.0    42188.0   
4   AUS03: Greater Brisbane    37580.0    37564.0    39080.0    40762.0   
5      AUS04: Greater Perth    45713.0    47371.0    48719.0    51020.0   
6   AUS05: Greater Adelaide    36505.0    37194.0    37634.0    37399.0   

   pcGDP2005  pcGDP2006  pcGDP2007  pcGDP2008  pcGDP2009  pcGDP2010  \
2    45423.0    45547.0    45880.0    45225.0    45900.0    45672.0   
3    41484.0    41589.0    42316.0    40975.0    41384.0    40943.0   
4    42976.0    44475.0    44635.0    46192.0    43507.0    42774.0   
5    55278.0    60142.0    62551.0    63899.0    63616.0    70111.0   
6    37604.0    38151.0    39049.0    38502.0    39538.0    39309.0   

   pcGDP2011  pcGDP2012  pcGDP2013  pcGDP2014  pcGDP2015  pcGDP2016  \
2    46535.0    47350.0    47225.0    48510.0    50075.0    50519.0   
3    41165.0    41264.0    41157.0    42114.0    42928.0    42671.0   
4    44166.0    43764.0    43379.0    43754.0    44388.0    45723.0   
5    73715.0    72679.0    76153.0    70395.0    66544.0    66032.0   
6    39223.0    39812.0    39855.0    40306.0    40295.0    39737.0   

   pcGDP2017  pcGDP2018  
2    50578.0    49860.0  
3    43025.0    42674.0  
4    46876.0    46640.0  
5    66424.0    70390.0  
6    40115.0    39924.0

In [19]:
percapitaGDP.shape

(480, 19)

Confirmamos que existem 480 valores válidos para a coluna `metro` e existem 480 únicos valores, antes de setar o index.

In [20]:
percapitaGDP.metro.count()

480

In [21]:
percapitaGDP.metro.nunique()

480

In [22]:
percapitaGDP.set_index('metro', inplace=True)

In [24]:
percapitaGDP.head()

pcGDP2001  pcGDP2002  pcGDP2003  pcGDP2004  \
metro                                                                  
AUS01: Greater Sydney       43313.0    44008.0    45424.0    45837.0   
AUS02: Greater Melbourne    40125.0    40894.0    41602.0    42188.0   
AUS03: Greater Brisbane     37580.0    37564.0    39080.0    40762.0   
AUS04: Greater Perth        45713.0    47371.0    48719.0    51020.0   
AUS05: Greater Adelaide     36505.0    37194.0    37634.0    37399.0   

                          pcGDP2005  pcGDP2006  pcGDP2007  pcGDP2008  \
metro                                                                  
AUS01: Greater Sydney       45423.0    45547.0    45880.0    45225.0   
AUS02: Greater Melbourne    41484.0    41589.0    42316.0    40975.0   
AUS03: Greater Brisbane     42976.0    44475.0    44635.0    46192.0   
AUS04: Greater Perth        55278.0    60142.0    62551.0    63899.0   
AUS05: Greater Adelaide     37604.0    38151.0    39049.0    38502.0   

                          pcGDP2009  pcGDP2010  pcGDP2011  pcGDP2012  \
metro                                                                  
AUS01: Greater Sydney       45900.0    45672.0    46535.0    47350.0   
AUS02: Greater Melbourne    41384.0    40943.0    41165.0    41264.0   
AUS03: Greater Brisbane     43507.0    42774.0    44166.0    43764.0   
AUS04: Greater Perth        63616.0    70111.0    73715.0    72679.0   
AUS05: Greater Adelaide     39538.0    39309.0    39223.0    39812.0   

                          pcGDP2013  pcGDP2014  pcGDP2015  pcGDP2016  \
metro                                                                  
AUS01: Greater Sydney       47225.0    48510.0    50075.0    50519.0   
AUS02: Greater Melbourne    41157.0    42114.0    42928.0    42671.0   
AUS03: Greater Brisbane     43379.0    43754.0    44388.0    45723.0   
AUS04: Greater Perth        76153.0    70395.0    66544.0    66032.0   
AUS05: Greater Adelaide     39855.0    40306.0    40295.0    39737.0   

                          pcGDP2017  pcGDP2018  
metro                                           
AUS01: Greater Sydney       50578.0    49860.0  
AUS02: Greater Melbourne    43025.0    42674.0  
AUS03: Greater Brisbane     46876.0    46640.0  
AUS04: Greater Perth        66424.0    70390.0  
AUS05: Greater Adelaide     40115.0    39924.0

In [25]:
percapitaGDP.loc['AUS02: Greater Melbourne']

pcGDP2001    40125.0
pcGDP2002    40894.0
pcGDP2003    41602.0
pcGDP2004    42188.0
pcGDP2005    41484.0
pcGDP2006    41589.0
pcGDP2007    42316.0
pcGDP2008    40975.0
pcGDP2009    41384.0
pcGDP2010    40943.0
pcGDP2011    41165.0
pcGDP2012    41264.0
pcGDP2013    41157.0
pcGDP2014    42114.0
pcGDP2015    42928.0
pcGDP2016    42671.0
pcGDP2017    43025.0
pcGDP2018    42674.0
Name: AUS02: Greater Melbourne, dtype: float64